text/0000-declarative-nonstandard-event-binding.md

```diff
@@ -0,0 +1,125 @@
+---
+title: Declarative Binding for Non-Standard Event Names
+status: DRAFTED
+created_at: 2024-02-25
+updated_at: 2024-02-25
+pr: (leave this empty until the PR is created)
+---
+
+# Declarative Binding for Non-Standard Event Names
+
+## Summary
+
+LWC support for declarative event binding has always been limited to event names using lower-case
+alphabetic characters, underscores, and numeric characters. Event names containing upper-case
+characters (e.g., camelCase, CAPScase, PascalCase, etc) have never been supported in LWC due to
+the case-insensitive nature of HTML. In addition, event names containing hyphens were never
+supported due to the reluctance to avoid introducing ambiguity into the LWC convention of using
+hyphens in attribute names to map to upper-case letters in property names.
+
+This has worked well in practice with developers sticking to LWC conventions and using
+`addEventListener()` when required as a workaround, but with the recent introduction of
+`lwc:external` which provides a paved path for integrating third party web components, there is
+greater value in providing a declarative way to listen for non-standard event names.
+
+This RFC proposes the introduction of a new template directive `lwc:on`, which allows us to support
+declarative event binding for third party web components, while sidestepping all the aforementioned
+limitations and ambiguities.
+
+## Basic example
+
+```js
+customElements.define('third-party', class extends HTMLElement{
+    connectedCallback() {
+        this.dispatchEvent(new CustomEvent('Foo-Bar-BAZ'));
+    }
+});
+```
+
+These are examples of some of the options that are currently on the table:
+
+```html
+<!-- Event as part of directive -->
+<third-party lwc:external lwc:on:Foo-Bar-BAZ={handleConnected}></third-party>
+
+<!-- Event as part of attribute name (special treatment for lwc:external) -->
+<third-party lwc:external onFoo-Bar-BAZ={handleConnected}></third-party>
+
+<!-- Event and listener as part of attribute value -->
+<third-party lwc:external lwc:on="Foo-Bar-BAZ:handleConnected"></third-party>
+```
+
+This is an example of a less declarative approach but should also be useful for things like creating
+components from metadata:
+
+```js
+// foo.js
+class Foo extends LightningElement {
+    handlers = {
+        'kebab-case': (event) => console.log("'kebab-case' handled"),
+        camelCase: (event) => console.log("'camelCase' handled"),
+        CAPScase: (event) => console.log("'CAPScase' handled"),
+        PascalCase: (event) => console.log("'PascalCase' handled"),
+    };
+}
+```

```

Comment:
This to me is unnecessary since we already have `lwc:spread`, which works with `onclick`/`onblur`/etc. We may have to consider how `lwc:spread` would interact, though, with `lwc:external` if we decided to support the non-standard event names there.